File: src/main/java/com/github/pauldambra/moduluschecker/chain/checks/ModulusElevenCheck.java
package com.github.pauldambra.moduluschecker.chain.checks;

import com.github.pauldambra.moduluschecker.Account.BankAccount;
import com.github.pauldambra.moduluschecker.ModulusCheckParams;
import com.github.pauldambra.moduluschecker.valacdosFile.ModulusElevenWeightsTransformer;
import com.github.pauldambra.moduluschecker.valacdosFile.WeightRow;
import com.google.common.collect.ImmutableList;

import java.util.function.Function;

public final class ModulusElevenCheck {

    public Boolean check(ModulusCheckParams params, Function<ModulusCheckParams, WeightRow> rowSelector) {
        WeightRow selectedRow = rowSelector.apply(params);

        ImmutableList<Integer> weights = ModulusElevenWeightsTransformer.CheckForExceptionTwoAndNine(params, selectedRow);

        int total = ModulusTotal.calculate(params.account, weights);
        int remainder = total % 11;

        if (selectedRow.isException(4)) {
            return remainder == exceptionFourCheckDigit(params);
        }
        else {
            boolean result = remainder == 0;
            if (result) { return true; }

            return new LloydsAlternateModulusElevenCheck().check(params, rowSelector);
        }
    }

    private int exceptionFourCheckDigit(ModulusCheckParams params) {
        int g = params.account.getNumberAt(BankAccount.G);
        int h = params.account.getNumberAt(BankAccount.H);
        return Integer.parseInt(String.format("%s%s",g,h));
    }

}

File: src/main/java/com/github/pauldambra/moduluschecker/chain/checks/ModulusElevenCheck.kt
package com.github.pauldambra.moduluschecker.chain.checks

import com.github.pauldambra.moduluschecker.ModulusCheckParams
import com.github.pauldambra.moduluschecker.account.BankAccount
import com.github.pauldambra.moduluschecker.valacdosFile.ModulusElevenWeightsTransformer
import com.github.pauldambra.moduluschecker.valacdosFile.WeightRow

class ModulusElevenCheck {

    fun check(params: ModulusCheckParams, weightRow: WeightRow): Boolean {

        val weights = ModulusElevenWeightsTransformer.checkForExceptionTwoAndNine(params, weightRow)

        val total = ModulusTotal.calculate(params.account, weights)
        val remainder = total % 11

        return if (weightRow.isException(4)) {
            remainder == exceptionFourCheckDigit(params)
        } else {
            if (remainder == 0) {
                true
            } else {
                LloydsAlternateModulusElevenCheck().check(params, weightRow)
            }

        }
    }

    private fun exceptionFourCheckDigit(params: ModulusCheckParams): Int {
        val g = params.account.getNumberAt(BankAccount.G)
        val h = params.account.getNumberAt(BankAccount.H)
        return Integer.parseInt(String.format("%s%s", g, h))
    }

}
